### PR TITLE
Add support of GroK command including default patterns

### LINKED ISSUE #451: Add PPL `Grok` command
Add PPL `grok` command described [here](https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/cmd/grok.rst)
Map `grok` to the next spark sql [`regexp_extract` command ](https://docs.databricks.com/en/sql/language-manual/functions/regexp_extract.html)

**Related campaign :**
https://github.com/opensearch-project/opensearch-spark/issues/408

### PR DESCRIPTION
**Pull request by YANG-DB**
### Description
Add PPL `grok` command described [here](https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/cmd/grok.rst)
Map `grok` to the next spark sql [`regexp_extract` command ](https://docs.databricks.com/en/sql/language-manual/functions/regexp_extract.html)

**Related campaign :**
https://github.com/opensearch-project/opensearch-spark/issues/408

### Issues Resolved
#451 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ppl-spark-integration/src/main/java/org/opensearch/sql/common/grok/GrokCompiler.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import java.time.ZoneId;
 import java.time.ZoneOffset;
 import java.util.HashMap;
+import java.util.List;
```
Comment (LantaoJin) [inline ppl-spark-integration/src/main/java/org/opensearch/sql/common/grok/GrokCompiler.java:22]: Useless code?

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/ParseStrategy.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        });
+        //list the group numbers of these projected fields
+        // match the regExpExtract group identifier with its number
+        namedGroupNumbers.forEach((group, index) -> {
+            //first create the regExp 
+            RegExpExtract regExpExtract = new RegExpExtract(sourceField,
+                    org.apache.spark.sql.catalyst.expressions.Literal.create(cleanedPattern, StringType),
+                    org.apache.spark.sql.catalyst.expressions.Literal.create(index + 1, IntegerType));
+            //next create Coalesce to handle potential null values 
+            Coalesce coalesce = new Coalesce(seq(regExpExtract));
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/utils/ParseStrategy.java:68]: that's a good idea

File: ppl-spark-integration/src/main/resources/patterns/firewalls [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Forked from https://github.com/elasticsearch/logstash/tree/v1.4.0/patterns
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply ppl-spark-integration/src/main/resources/patterns/firewalls:1]: yes - its for future usage in the `patterns` ppl command

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java [hunk 4]
```diff
@@ -387,7 +361,7 @@ public LogicalPlan visitDedupe(Dedupe node, CatalystPlanContext context) {
         visitFieldList(node.getFields(), context);
         // Columns to deduplicate
         Seq<org.apache.spark.sql.catalyst.expressions.Attribute> dedupFields
-            = context.retainAllNamedParseExpressions(e -> (org.apache.spark.sql.catalyst.expressions.Attribute) e);
+                = context.retainAllNamedParseExpressions(e -> (org.apache.spark.sql.catalyst.expressions.Attribute) e);
```
[the remark this answers is not in the document]
Comment (YANG-DB) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:354]: good point..

File: ppl-spark-integration/README.md [hunk 5]
```diff
@@ -314,6 +314,14 @@ Limitation: Overriding existing field is unsupported, following queries throw ex
...
 - `source=accounts | parse address '(?<streetNumber>\d+) (?<street>.+)' | where streetNumber > 500 | sort num(streetNumber) | fields streetNumber, street`
 
+**Grok**
+- `source=accounts | grok email '.+@%{HOSTNAME:host}' | top 1 host`
+- `source=accounts | grok email '.+@%{HOSTNAME:host}' | stats count() by host`
+- `source=accounts | grok email '.+@%{HOSTNAME:host}' | eval eval_result=1 | fields host, eval_result`
+- `source=accounts | grok email '.+@%{HOSTNAME:host}' | eval eval_result=1 | fields host, eval_result`
+- `source=accounts | grok street_address '%{NUMBER} %{GREEDYDATA:address}' | fields address `
+- `source=logs | grok message '%{COMMONAPACHELOG}' | fields COMMONAPACHELOG, timestamp, response, bytes`
+-
```
Comment (LantaoJin) [inline ppl-spark-integration/README.md:324]: L324 need to remove.
BTW, can we add a line to explain current limitation of Grok? such as 
```
Limitation: Overriding existing field is unsupported:
accounts | grok address '%{NUMBER} %{GREEDYDATA:address}' | fields address
```

Comment (LantaoJin) [reply ppl-spark-integration/README.md:324]: LGTM for rest

**Review by LantaoJin — APPROVED** [review]

**Pull request merged**